マイプランページの実装

totota-08 opened the pull request:
## 対応したこと

- recent_planのタグと画像を引数で受け取れるように変更
- home_pageなどでrecent_planが横幅を突き破る問題を修正
- マイプランページのブックマークページの実装
- マイプランページの作成したプランページの実装

## 画面キャプチャ(全てAndroidによる実機ビルド)
### 最近作成したプランの修正前と修正後
<p style="display: flex; justify-content: space-around;">
  <img src="https://github.com/user-attachments/assets/d539f279-2e3c-484a-944b-178edc964988" alt="Image 1" width="45%">
  <img src="https://github.com/user-attachments/assets/d89e51fa-5b95-43b7-95d4-d74c1af1c04a" alt="Image 2" width="45%">
</p>

### ブックマークページ
<p style="display: flex; justify-content: space-around; align-items: center;">
  <img src="https://github.com/user-attachments/assets/899537bc-a165-4217-9ddf-fbca91231e6e" alt="Image 1" width="45%">
 <img src="https://github.com/user-attachments/assets/5c1deb2f-7f80-4f36-a7b3-cda03f5ef1a5" alt="Image 1" width="45%">
</p>

### データがない場合のボタンの動作
<p style="display: flex; justify-content: space-around; align-items: center;">
  <img src="https://github.com/user-attachments/assets/74013679-b4ad-4594-9f75-983571dd236e" alt="Image 1" width="45%">
 <img src="https://github.com/user-attachments/assets/d7689305-5a0f-4086-bb7a-e566c6763456" alt="Image 1" width="45%">
</p>

### エラーページ
<p style="display: flex; justify-content: space-around;">
  <img src="https://github.com/user-attachments/assets/8ac9f12d-76bb-4b71-8efd-62a3222167e7" alt="Image 1" width="45%">
  <img src="https://github.com/user-attachments/assets/63af9aa1-3384-4ccd-be5b-432e46494904" alt="Image 2" width="45%">
</p>

totota-08 committed:
画像urlとタグを引数で受け取れるように変更

totota-08 committed:
画像urlとタグを引数で受け取れるように変更

totota-08 committed:
データを受け取った後にリストで返すウィジェットの作成

totota-08 committed:
データを受け取った後にリストで返すウィジェットの作成

totota-08 committed:
ダミーの旅行データ

totota-08 committed:
ダミーの旅行データ

totota-08 committed:
Revert "ダミーの旅行データ"

This reverts commit c7a861811323b73daf055e0b14fcb70c2a39c3aa.

totota-08 committed:
Revert "ダミーの旅行データ"

This reverts commit c7a861811323b73daf055e0b14fcb70c2a39c3aa.

totota-08 committed:
自動生成

totota-08 committed:
タグの外側にSizedBoxを配置

totota-08 committed:
ダミーデータの作成

totota-08 committed:
ユーザーデータの取得

totota-08 committed:
横幅専用のCategorytagsの作成。

既存のCategorytagsがwarpで定義されていたのをRowに変更
タグの大きさをテキストの横幅から丁寧に計算するメソッドの追加

totota-08 committed:
RecentPlanをリストで返すウィジェットの作成

Planのリストを受け取りRecentPlanが入ったリストで返す。
Columnなどで使う可能性を踏まえListViewでは返していない。

totota-08 committed:
リスト名メソッド名の変更、thumbnailUrlの変更

totota-08 committed:
削除

totota-08 committed:
RowCategoryTagsの追加

totota-08 committed:
PlanListを使い表示するように変更

totota-08 committed:
PlanListのPaddingが変わったため変更

totota-08 committed:
ブックマーク付きのプラン一覧画面の実装

totota-08 committed:
データが一つもない時に表示するウィジェット（仮）

totota-08 committed:
デザイン通りに変更

totota-08 committed:
Firebaseでユーザーのブックマークしているプランの情報を取得する機能の実装

totota-08 committed:
ユーザーがブックマークしているリストを状態管理

totota-08 committed:
asyncStateを使いブックマークしたプランを表示するよう変更

totota-08 committed:
文字色を変更

totota-08 committed:
自動生成

totota-08 committed:
Merge branch 'develop' into feature/my-plan-tabs

# Conflicts:
#	lib/presentation/home/components/recent_plan.dart
#	lib/presentation/home/components/recent_plans_section.dart
#	lib/presentation/my_plan/bookmark_plans_tab_view.dart
#	lib/presentation/my_plan/created_plans_tab_view.dart

totota-08 committed:
ブックマークを解除した時の処理を追加

totota-08 committed:
ブックマークボタンの機能を実装

totota-08 committed:
多言語対応

totota-08 committed:
自動生成

totota-08 committed:
firebaseの処理内容を記述

totota-08 committed:
buildEllipsisTagウィジェットの長さを足しオーバーフロー対策

totota-08 committed:
不要なColumnを除去

totota-08 committed:
ボタンを追加多言語対応

totota-08 committed:
エラー発生時に別のページに遷移する機能の実装

nonndataウィジェットのListViewを削除

totota-08 committed:
作成したプランの状態を管理するリストの追加

totota-08 committed:
firebaseの処理をDataSourceに移行

totota-08 committed:
エラーページを表示するよう変更

totota-08 committed:
自動生成

totota-08 committed:
エラー文章の多言語対応

totota-08 committed:
遷移先を変更

totota-08 committed:
analyzeのエラーが出ないよう修正

Haru-Kobayashi073 commented on lib/infrastructure/bookmark/bookmark_data_sorce.dart:
```diff
@@ -0,0 +1,79 @@
+import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+import '../../i18n/strings.g.dart';
```
新しくDataSourceのファイルを作成する時は、
1. 基本対応するモデル(Plan, Place, Userなど(例外: Gemini, File))に対して作成する
2. Planのブックマークの切り替えなので、PlanDataSourceクラスに定義する(クラスはすでに定義ずみ)

ので、今回ブックマークはモデルではなく、行為を表すものなので、DataSourceとして個別に定義はしなくても大丈夫
PlanDataSourceにまとめよう

Haru-Kobayashi073 replied:
PlanDataSourceクラスなどのプラン作成のメソッドなどをみてもらうとわかると思うけれど、エラーハンドリングは行わないです。
基本やるのは、

1. firestoreから素のデータを受け取る。
2. プロジェクト内で定義してあるモデル(Plan, Place, Userなど)にシリアライズする

のみです
Map<String, dynamic>のような扱いずらい状態でnotifierに渡すのはやめましょう
エラーハンドリングは、notifierで行いましょう
俺が今対応してる部分で、マージはまだだけど、[このファイル](https://github.com/Haru-Kobayashi073/Shibuya-Buddy/pull/111/files#diff-b8669c7a838398095621832c7d178f27b14c024c4eec93a66673b91e16992be4R45-R51)が参考になるはず！

Haru-Kobayashi073 commented on lib/infrastructure/bookmark/bookmark_data_sorce.dart:
```diff
@@ -0,0 +1,79 @@
+  FirebaseFirestore get firestore => ref.read(cloudFirestoreProvider);
+  ScaffoldMessenger get scaffoldMessenger =>
+      ref.read(scaffoldMessengerProvider.notifier);
+  BookmarkDataSorce get booksorce =>
```
消す予定のファイルだと思うので、あまりレビューしてもかもだけど、
あるクラス(今回だとBookmarkDataSource)を定義している中で、定義元のクラスのゲッターを定義することはほぼないよ

Haru-Kobayashi073 commented on lib/infrastructure/bookmark/bookmark_data_sorce.dart:
```diff
@@ -0,0 +1,79 @@
+@riverpod
+class BookmarkDataSorce extends _$BookmarkDataSorce {
+  FirebaseFirestore get firestore => ref.read(cloudFirestoreProvider);
+  ScaffoldMessenger get scaffoldMessenger =>
```
他のコメントでも言ってるけれど、エラーハンドリングはこのファイルではしないので、Scaffold Messengerを呼び出すことはないよ

Haru-Kobayashi073 commented on lib/presentation/my_plan/my_plan_state.dart:
```diff
@@ -0,0 +1,14 @@
+part 'my_plan_state.freezed.dart';
+
+@freezed
+class MyPlanState with _$MyPlanState {
```
細かいですが、他に統一するとMyPlanPageStateクラスだね

Haru-Kobayashi073 commented on lib/presentation/my_plan/my_plan_notifier.dart:
```diff
@@ -0,0 +1,86 @@
+
+  custom.ScaffoldMessenger get scaffoldMessenger =>
+      ref.read(scaffoldMessengerProvider.notifier);
+  BookmarkDataSorce get booksorce =>
```
BookmarkDataSourceクラスは今回PlanDataSourceクラスに統合されるべきなので、必要ないですが、DataSource層の呼び出し方は合ってます！

Haru-Kobayashi073 commented on lib/presentation/my_plan/my_plan_notifier.dart:
```diff
@@ -0,0 +1,86 @@
+
+@riverpod
+class MyPlanNotifier extends _$MyPlanNotifier {
+  FirebaseFirestore get firestore => ref.read(cloudFirestoreProvider);
```
firestoreのインスタンスを扱うようなコアの処理はnotifierでは絶対に書いてはいけないので、このfirestoreゲッターを使わなくても良くなるようなnotifierの書き方になると最高

Haru-Kobayashi073 commented on lib/presentation/my_plan/my_plan_notifier.dart:
```diff
@@ -0,0 +1,86 @@
+part 'my_plan_notifier.g.dart';
+
+@riverpod
+class MyPlanNotifier extends _$MyPlanNotifier {
```
細かいけれど、他のクラスの書き方に合わせるとMyPlanPageNotifierになる

Haru-Kobayashi073 commented on lib/presentation/my_plan/my_plan_notifier.dart:
```diff
@@ -0,0 +1,86 @@
+    try {
+      for (final id in planIds) {
+        final data = await booksorce.getPlanData(planId: id);
+        for (final topic in data['topics'] as List<dynamic>) {
```
`as ~~~`のようなものを型のキャストと言います。
いわゆる言語に〜〜〜型だよと無理やり理解させるような危険な行為です。
いつか違う型が入ってきてしまった時にエラーで止まってしまいます。
なので、型のキャストは行わずに、元々ぴったりのTopicというクラスが用意されているので、そっちを使おう

Haru-Kobayashi073 commented on lib/presentation/my_plan/created_plans_tab_view.dart:
```diff
@@ -1,14 +1,57 @@
+          },
+        );
+      },
+      loading: () {
+        return const Center(child: CircularProgressIndicator());
+      },
```
ローディングは共通化済みなので、以下のように短縮して書けます
Loadingクラスはしっかりインポートしてください

```suggestion
      loading: Loading.new,
```

Haru-Kobayashi073 reviewed (changes requested):
@PenginPenPen 
一旦現状の修正お願い！
notifierとDataSourceは実装ずみのファイルを見て真似して欲しい！🙏🙏🙏
わからなかったらすぐ聞いてくれて大丈夫

totota-08 committed:
plan_data_sourceに移動

totota-08 committed:
topicのidを受け取りTopicで返すtopic_data_sourceを実装

totota-08 committed:
メソッドの名前を変更

totota-08 committed:
メソッドの命名変更と不要なメソッドを削除

totota-08 committed:
自動生成

totota-08 committed:
dart format

Haru-Kobayashi073 commented on lib/presentation/my_plan/created_plans_tab_view.dart:
```diff
@@ -1,14 +1,51 @@
+                      child: PlanList(plans: plans),
+                    ),
+                  ],
+                ),
```
PlanListがListViewウィジェットで囲われているはずなので、一番外側のListViewウィジェットは必要ないはず
```suggestion
                child: Padding(
                      padding: const EdgeInsets.all(16),
                      child: PlanList(plans: plans),
                    ),
```

Haru-Kobayashi073 commented on lib/presentation/my_plan/components/nondata.dart:
```diff
@@ -0,0 +1,44 @@
+import '../../../utils/styles/app_text_style.dart';
+import '../../components/wide_button.dart';
+
+class Nondata extends StatelessWidget {
```
名前が少しわかりづらいような気がするので、NonDataViewやEmptyDataViewなどにしましょう

```suggestion
class EmptyDataView extends StatelessWidget {
```

Haru-Kobayashi073 commented on lib/presentation/my_plan/my_plan_state.dart:
```diff
@@ -0,0 +1,14 @@
+import 'package:flutter/foundation.dart';
```
ファイル名も合わせて改名しましょう

Haru-Kobayashi073 commented on lib/infrastructure/plan/plan_data_source.dart:
```diff
@@ -40,4 +46,59 @@ class PlanDataSource extends _$PlanDataSource implements PlanRepository {
+      final topic = await topicsorce.getTopicData(topicId: topicId.toString());
+      topics.add(Topic(name: topic.name, thumbnailUrl: ''));
+    }
+    return Plan(
+      id: planId,
+      title: data['title'].toString(),
+      description: data['description'].toString(),
+      thumbnailUrl: data['thumbnailUrl'].toString(),
+      topics: topics,
+    );
+  }
```
ここも他のメソッドの書き方を真似しましょう
直接Map<String, dynamic>の型から文字列でデータを抽出している部分はないはず
fromJsonメソッドというfirestoreから受け取ったデータをDartの型に直す便利なメソッドがあるので、使いましょう

```suggestion
    final res = Plan.fromJson(doc.data());
    return res.copyWith(
      topics: topics,
    );
  }
```

Haru-Kobayashi073 commented on lib/infrastructure/topic/topic_data_source.dart:
```diff
@@ -0,0 +1,36 @@
+part 'topic_data_source.g.dart';
+
+@riverpod
+class TopicDataSource extends _$TopicDataSource {
```
他のDataSourceクラスを真似しましょう！！
implementsでRepositoryクラスを継承しているはずです

Haru-Kobayashi073 commented on lib/infrastructure/topic/topic_data_source.dart:
```diff
@@ -0,0 +1,36 @@
+      return const Topic(name: '', thumbnailUrl: 'thumbnailUrl');
+    }
+    final ranking = int.tryParse(topicData['ranking'].toString()) ?? 0;
+    return Topic(
```
ここもfromJsonメソッドを使いましょう

Haru-Kobayashi073 reviewed (changes requested):
一旦これで第二弾にしよう

totota-08 committed:
Merge branch 'develop' into feature/my-plan-tabs

# Conflicts:
#	lib/i18n/strings.g.dart
#	lib/i18n/strings_en.g.dart
#	lib/i18n/strings_ja.g.dart
#	lib/i18n/strings_zh_Hans.g.dart
#	lib/i18n/strings_zh_Hant.g.dart

totota-08 committed:
自動生成

totota-08 committed:
ファイル名を変更

totota-08 committed:
ファイル名を変更

totota-08 committed:
ウィジェット名を変更

totota-08 committed:
Planlistの外側ListViewを削除

totota-08 committed:
不要な処理を削除

totota-08 committed:
topicDataSourceのリポジトリを作成

totota-08 committed:
ファイル名の変更

Haru-Kobayashi073 commented on lib/presentation/my_plan/bookmark_plans_tab_view.dart:
```diff
@@ -1,14 +1,56 @@
+                  await notifier.buildBookmarkPlans();
+                },
+                child: ListView(
+                  children: [
```
BookmarkPlanListウィジェットが既にListViewになってると思うので、1番外側のListViewは必要ないかも

Haru-Kobayashi073 reviewed (changes requested):
修正お願いします！
俺が実装してる部分と割と被ってる部分もあるから、書き換えとか起こったらごめん🙏🏻

totota-08 committed:
リストビューを削除

totota-08 committed:
Planに渡すデータを変更

Plan.fromJsonでそのままdataを渡すとtopicが型の不一致でエラーになってしまうため、一回mapオブジェクトに変換後渡している。bookmarkを削除する時にidが必要なので、copyWithでidを受け渡している。

totota-08 committed:
ColumnではなくListViewで返すように変更

totota-08 committed:
自動生成

Haru-Kobayashi073 commented on lib/infrastructure/plan/plan_data_source.dart:
```diff
@@ -40,4 +46,58 @@ class PlanDataSource extends _$PlanDataSource implements PlanRepository {
+    final snapshot =
+        await firestore.collection('PopularPlans').doc(planId).get();
+    final data = snapshot.data();
+    if (data == null) {
```
基本的にdataが空かどうかのエラーハンドリングはNotifierに任せたいので、!を使って強制NonNullにしたりで対応して欲しいかも
逆に空のPlanデータを返すと、notifier側でせっかく書いたエラーハンドリングが発動しないから

Haru-Kobayashi073 replied:
DataSource系で他の箇所で空かの確認をしている箇所は合わせて欲しい

Haru-Kobayashi073 commented on lib/infrastructure/plan/plan_data_source.dart:
```diff
@@ -40,4 +46,58 @@ class PlanDataSource extends _$PlanDataSource implements PlanRepository {
+      );
+    }
+    for (final topicId in data['topics'] as List) {
+      final topic = await topicsorce.getTopicData(topicId: topicId.toString());
+      topics.add(Topic(name: topic.name, thumbnailUrl: ''));
```
getTopicDataメソッドを使用するとそのままTopic型のデータが返ってくるので、そのまま追加しよう
```suggestion
      final topic = await topicsorce.getTopicData(topicId: topicId.toString());
      topics.add(topic);
```

Haru-Kobayashi073 commented on lib/infrastructure/plan/plan_data_source.dart:
```diff
@@ -40,4 +46,58 @@ class PlanDataSource extends _$PlanDataSource implements PlanRepository {
+      final topic = await topicsorce.getTopicData(topicId: topicId.toString());
+      topics.add(Topic(name: topic.name, thumbnailUrl: ''));
+    }
+    final topicMaps = topics.map((topic) => topic.toJson()).toList();
+    final processedData = Map<String, dynamic>.from(data);
+    processedData['topics'] = topicMaps;
+
+    final res = Plan.fromJson(processedData);
+    return res.copyWith(id: planId);
```
List<Topic>型のデータをそのまま代入しましょう
```suggestion
    final res = Plan.fromJson(data);
    return res.copyWith(
        id: planId,
        topics: topics,
        );
```

Haru-Kobayashi073 commented on lib/infrastructure/topic/topic_data_source.dart:
```diff
@@ -0,0 +1,32 @@
+    if (topicData == null) {
+      return const Topic(name: '', thumbnailUrl: 'thumbnailUrl');
+    }
+    final ranking = int.tryParse(topicData['ranking'].toString()) ?? 0;
+    final ref = Topic.fromJson(topicData);
+    return ref.copyWith(ranking: ranking);
```
rankingフィールドは空の場合を許容するデータ型(int? ranking)にしてあるので、条件分岐はしなくても大丈夫
```suggestion
    return Topic.fromJson(topicData);
```

Haru-Kobayashi073 commented on lib/presentation/my_plan/components/bookmark_plan_list.dart:
```diff
@@ -0,0 +1,57 @@
+
+  final List<Plan> plans;
+
+  List<Widget> _buildPlanItems(BuildContext context, WidgetRef ref) {
```
PlanListクラスと同様にこのくらいのコード量なら、メソッドを分けずに直接書こう

Haru-Kobayashi073 commented on lib/presentation/home/components/plan_list.dart:
```diff
@@ -0,0 +1,34 @@
+import '../../../domain/entities/plan.dart';
+import 'recent_plan.dart';
+
+class PlanList extends StatelessWidget {
```
ほぼまったく同じコンポーネントをmy_planフォルダ内にも作っているので、このファイルをpresentation/components/配下に移動しましょう

Haru-Kobayashi073 commented on lib/presentation/home/components/plan_list.dart:
```diff
@@ -0,0 +1,34 @@
+
+  final List<Plan> plans;
+
+  List<Widget> _buildPlanItems() {
+    return plans
+        .map(
+          (plan) => Padding(
+            padding: const EdgeInsets.only(bottom: 8),
+            child: RecentPlan(
+              title: plan.title,
+              imageUrl: plan.thumbnailUrl,
+              tags: plan.topics.map((e) => e.name).toList(),
+            ),
+          ),
+        )
+        .toList();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      children: _buildPlanItems(),
+    );
+  }
```
このくらいのコード量なら、メソッドは分けずに直接書いちゃおう

```suggestion
  @override
  Widget build(BuildContext context) {
    return Column(
      children: plans
        .map(
          (plan) => Padding(
            padding: const EdgeInsets.only(bottom: 8),
            child: RecentPlan(
              title: plan.title,
              imageUrl: plan.thumbnailUrl,
              tags: plan.topics.map((e) => e.name).toList(),
            ),
          ),
        ).toList(),
    );
  }
```

Haru-Kobayashi073 commented on lib/presentation/my_plan/components/plan_list.dart:
```diff
@@ -0,0 +1,33 @@
+import 'package:flutter/material.dart';
```
home/components/配下のPlanListクラスをpresentation/components/配下に移動をして、そのコンポーネントを使えばいいので、このファイルは消しちゃおう(中身がほぼ一緒なので)

Haru-Kobayashi073 reviewed (changes requested):
お願いします！

totota-08 committed:
homeにあるコンポーネントを移動

totota-08 committed:
不要なメソッドを削除し、ページに直接記述

totota-08 committed:
コンポーネントの内容を記述

totota-08 committed:
nullチェックを削除

totota-08 committed:
コンポーネントが移動したことにより変更

totota-08 committed:
Merge branch 'develop' into feature/my-plan-tabs

# Conflicts:
#	lib/domain/repositories/topic_repository.dart
#	lib/infrastructure/plan/plan_data_source.dart
#	lib/infrastructure/plan/plan_data_source.g.dart
#	lib/infrastructure/topic/topic_data_source.dart
#	lib/infrastructure/topic/topic_data_source.g.dart
#	lib/presentation/home/components/recent_plan.dart
#	lib/presentation/home/components/recent_plans_section.dart

totota-08 committed:
自動生成

totota-08 committed:
import文を修正

Haru-Kobayashi073 commented on lib/infrastructure/plan/plan_data_source.dart:
```diff
@@ -69,4 +75,42 @@ class PlanDataSource extends _$PlanDataSource implements PlanRepository {
     return snapshot.docs.map((doc) => Plan.fromJson(doc.data())).toList();
   }
+
+  Future<List<String>> getBookmarkedPlanIds() async {
```
PlanRepositoryを継承しているので、すべてのメソッドに`@override`が付くはず。
追加してくれたメソッドはすべてついてないので、PlanRepositoryに追加したいメソッドを追加して欲しい

Haru-Kobayashi073 commented on lib/infrastructure/topic/topic_data_source.dart:
```diff
@@ -16,6 +16,13 @@ class TopicDataSource extends _$TopicDataSource implements TopicRepository {
   }
 
+  @override
+  Future<Topic> getTopicData({required String topicId}) async {
```
このシンプルさが完璧！

Haru-Kobayashi073 commented on lib/presentation/components/plan_list.dart:
```diff
@@ -0,0 +1,33 @@
+
+  final List<Plan> plans;
+
+  List<Widget> _buildPlanItems() {
+    return plans
+        .map(
+          (plan) => Padding(
+            padding: const EdgeInsets.only(bottom: 8),
+            child: RecentPlan(
+              plan: plan,
+            ),
+          ),
+        )
+        .toList();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      children: _buildPlanItems(),
+    );
+  }
```
**再喝**
このくらいのコード量なら、メソッドは分けずに直接書いちゃおう


```suggestion
  @override
  Widget build(BuildContext context) {
    return Column(
      children: plans
        .map(
          (plan) => Padding(
            padding: const EdgeInsets.only(bottom: 8),
            child: RecentPlan(
              plan: plan,
            ),
          ),
        )
        .toList(),
    );
  }
```

Haru-Kobayashi073 reviewed (changes requested):
再喝のコメントが多いですが、修正お願いします
これ終われば、LGTMなる！

totota-08 committed:
メソッドを追加

totota-08 committed:
PopularPlansではなくplansからデータを取得

totota-08 committed:
メソッドにわけていたのを統合

Haru-Kobayashi073 reviewed (approved):
今後不具合見つけ次第書き換えるかもだけど、一旦LGTM！
